compiler/src/lnntoamm/Types.ts

```diff
@@ -278,84 +550,127 @@ export class FunctionType extends Type {
-      let weightFns = fnsByWeight.get(weight);
-      weightFns = weightFns.filter(
-        ([weightedFn, _retTy]) =>
-          fns.findIndex(([fn, _retTy]) => fn === weightedFn) === -1,
-      );
-      return [...fns, ...weightFns];
-    }, new Array<[Fn, Type]>());
-    if (ret.length > originalLength) {
+    // const ret = weights.reduce((fns, weight) => {
+    // let weightFns = fnsByWeight.get(weight);
+    // isDbg && console.log('at weight', weight, 'fns are', weightFns);
+    // weightFns.forEach(([fn, pTys, retTy]) => {
+    //   let foundIdx = fns.findIndex(([maybeSame, _p, _r]) => maybeSame === fn);
+    //   if (foundIdx !== -1) {
+    //     const [[_sameFn, samePTys, sameRetTys]] = fns.splice(foundIdx, 1);
+    //   } else {
+    //     fns.push([fn, pTys, retTy]);
+    //   }
+    // });
+    // weightFns = weightFns.filter(
+    //   ([weightedFn, _retTy]) =>
+    //     fns.findIndex(([fn, _retTy]) => fn === weightedFn) === -1,
+    // );
+    // isDbg && console.log('filtered:', weightFns);
+    // return [...fns, ...weightFns];
+    const ret: [Fn[], Type[][], Type[]] = weights.reduce(
+      ([fns, pTys, retTys], weight) => {
+        fnsByWeight
+          .get(weight)
+          .forEach(([weightedFn, weightedPTys, weightedRetTy]) => {
+            const alreadyIdx = fns.findIndex((fn) => fn === weightedFn);
+            if (alreadyIdx === -1) {
+              // push to the end of the fns since technically the weight for the function
+              // is indeed higher than what it was before
+              fns.push(weightedFn);
+            } else {
+              fns.push(fns.splice(alreadyIdx, 1)[0]);
+            }
+            pTys = weightedPTys.map((pTy, ii) => [...(pTys[ii] || []), pTy]);
+            retTys.push(weightedRetTy);
+          });
+        return [fns, pTys, retTys];
+      },
+      [new Array<Fn>(), new Array<Type[]>(), new Array<Type>()] as [
+        Fn[],
+        Type[][],
+        Type[],
+      ],
+    );
+    if (ret[0].length > original.length || ret[0].length === 0) {
       console.log('~~~ ERROR');
-      console.log('original: ', originalLength);
-      console.log('retLength:', ret.length);
+      console.log('original: ', original);
+      console.log('ret:      ', ret);
+      console.log('retLength:', ret[0].length);
       console.log('args:     ', args);
       console.log('matrix:   ', matrix);
       console.log('byweight: ', fnsByWeight);
-      console.log('indices:  ', indices);
-      throw new Error('somehow got more options when fn selecting');
+      if (ret[0].length === 0) {
+        throw new Error('no more functions left');
+      } else {
+        throw new Error('somehow got more options when fn selecting');
+      }
     }
+    // const getStack = { stack: '' };
+    // Error.captureStackTrace(getStack);
+    isDbg &&
+      (() => {
+        //   console.log('returning from', getStack.stack);
+        console.dir(ret, { depth: 4 });
```

Comment:
Why is this in an IIFE?

Reply:
because otherwise i'd have to write a bunch of `isDbg && ...` when debugging